mycroft/skills/common_query_skill.py

```diff
@@ -80,39 +105,76 @@ def __handle_question_query(self, message):
             level = result[1]
             answer = result[2]
             callback = result[3] if len(result) > 3 else None
-            confidence = self.__calc_confidence(match, search_phrase, level)
-            self.bus.emit(message.response({"phrase": search_phrase,
-                                            "skill_id": self.skill_id,
-                                            "answer": answer,
-                                            "callback_data": callback,
-                                            "conf": confidence}))
+            confidence = self.__calc_confidence(match, search_phrase, level, answer)
+            self.bus.emit(
+                message.response(
+                    {
+                        "phrase": search_phrase,
+                        "skill_id": self.skill_id,
+                        "answer": answer,
+                        "callback_data": callback,
+                        "conf": confidence,
+                    }
+                )
+            )
         else:
             # Signal we are done (can't handle it)
-            self.bus.emit(message.response({"phrase": search_phrase,
-                                            "skill_id": self.skill_id,
-                                            "searching": False}))
-
-    def __calc_confidence(self, match, phrase, level):
+            self.bus.emit(
+                message.response(
+                    {
+                        "phrase": search_phrase,
+                        "skill_id": self.skill_id,
+                        "searching": False,
+                    }
+                )
+            )
+
+    def remove_noise(self, phrase):
+        # remove noise to produce essence
+        phrase = " " + phrase + " "
+        for word in self.translated_noise_words:
+            mtch = " " + word + " "
+            if phrase.find(mtch) > -1:
+                phrase = phrase.replace(word, "")
```

Comment:
consider porting the remove_voc method from OVOS

https://github.com/OpenVoiceOS/OVOS-workshop/blob/master/ovos_workshop/patches/base_skill.py#L165

Reply:
I will take a look at it your code J, thanks. FYI in a subsequent version of the code which I have not pushed I have a simple relevance = len(set(answer.split()) - set(noise_words)) / len(answer.split() but I felt this loop was a bit easier for folks to understand.

Reply:
Your code says # PR not yet released, and I have nod idea what voc file you are using, however, I thought I would simply post this to show that we are talking about 6 one way, half dozen another. 


class test():
    def __init__(self):
        self.translated_noise_words = ['where',
'which',
'they',
'them',
'when',
'what',
'that',
'will',
'from',
'that',
'also',
'who',
'how',
'did',
'and',
'but',
'the',
'too',
'why',
'is',
'it',
'do',
'or',
'to',
'of',
'a']

    def remove_noise_alt1(self, utt):
        for i in self.translated_noise_words:
            utt = re.sub(r'\b' + i + r"\b", "", utt)

        return utt

    def remove_noise_dont_preserve_order(self, phrase):
        # remove noise to produce essence
        phrase = " " + phrase + " "
        result_set = set( set(phrase.split(' ')) - set(self.translated_noise_words) )
        return (' '.join(result_set)).strip()

    def remove_noise_preserve_order(self, phrase):
        # remove noise to produce essence
        phrase = " " + phrase + " "
        for word in self.translated_noise_words:
            mtch = " " + word + " "
            if phrase.find(mtch) > -1:
                phrase = phrase.replace(word, "")
        phrase = " ".join(phrase.split())
        return phrase.strip()

    def __calc_confidence(self, match, phrase, level, answer):
        # Assume the more of the words that get consumed, the better the match
        consumed_pct = len(match.split()) / len(phrase.split())
        if consumed_pct > 1.0:
            consumed_pct = 1.0
        consumed_pct /= 10

        bonus = 0

        # extract topic
        topic = self.remove_noise(match)

        # calculate relevance
        matches = 0
        for word in topic:
            if answer.find(word) > -1:
                matches += 1

        answer_size = len(answer.split(" "))
        relevance = 0.0
        if answer_size > 0:
            relevance = float(float(matches) / float(answer_size))

        # extra credit for more words
        wc_mod = float(float(answer_size) / float(100))

        confidence = 0.6

        return confidence

phrases = [
        "is the empre state building the tallest building in the world",
        "which species of hummingbird is the smallest",
        "why did they burn joan of ark at the stake",
        "how tall is the eifel tower",
        "how deep is the marriana trench",
        "how old is paul mccartney"
        ]
t = test()

print("Order Preserved")
for phrase in phrases:
    topic = t.remove_noise_preserve_order(phrase)
    print("%s --- %s" % (topic, phrase))

print('')
print("Order Not Preserved")
for phrase in phrases:
    topic = t.remove_noise_dont_preserve_order(phrase)
    print("%s --- %s" % (topic, phrase))

print('')
print("Alt 1")
for phrase in phrases:
    topic = t.remove_noise_alt1(phrase)
    print("%s --- %s" % (topic, phrase))

Which produces 

species hummingbird smallest --- which species of hummingbird is the smallest
burn joan ark at stake --- why did they burn joan of ark at the stake
tall eifel tower --- how tall is the eifel tower
deep marriana trench --- how deep is the marriana trench
old paul mccartney --- how old is paul mccartney

Order Not Preserved
building state empre in world tallest --- is the empre state building the tallest building in the world
smallest species hummingbird --- which species of hummingbird is the smallest
stake ark at burn joan --- why did they burn joan of ark at the stake
tall eifel tower --- how tall is the eifel tower
trench deep marriana --- how deep is the marriana trench
paul mccartney old --- how old is paul mccartney

Alt 1
  empre state building  tallest building in  world --- is the empre state building the tallest building in the world
 species  hummingbird   smallest --- which species of hummingbird is the smallest
   burn joan  ark at  stake --- why did they burn joan of ark at the stake
 tall   eifel tower --- how tall is the eifel tower
 deep   marriana trench --- how deep is the marriana trench
 old  paul mccartney --- how old is paul mccartney

Reply:
So guys, do you really need me to change this to improve the PR?

Reply:
the PR not made comment in that link means i intended to submit that util method to the base MycroftSkill class, i havent made that PR yet.

Reply:
you would need to add the `noise.voc` file to `mycroft/res/text/en-us` and then 

```python
clean_utt = self.remove_voc(utterance, "noise.voc")
```

Reply:
Oh, and BTW, from a timing perspective ...

For 10,000 iterations on my laptop

Order preserved took 0.80367398262
Not Order preserved took 0.282736063004
Alt1 took 3.85695791245

Unsurprisingly sets are faster but loops are probably easier to read.

Reply:
would be curious what the timing would be with my approach, be sure to also post it if you measure it

Reply:
My issue with the implementation in the PR is that it malforms the output

this is an minimal example constructed in the repl using the PR branch:

```python
>> class T(CommonQuerySkill):
...   def __init__(self):
...     pass
...   def CQS_match_query_phrase(self, phrase):
...     pass
...
>>> t = T()
>>> t.translated_noise_words = ['is', 'and']
>>> t.remove_noise("this is andy and he is cool")
'th y he cool'
```

since words in the output is used later I assume it needs to not do partial removes.

But if that is not an issue this version is fine.

Reply:
See above output from my previous comment. It does not behave as your example does. The only issue is whether order is preserved and that is not important in this case.

Performance numbers were posted above but just to verify. 

For 10,000 iterations on my laptop

Order preserved took 0.80367398262
Not Order preserved took 0.282736063004
Alt1 took 3.85695791245

So loop easier to understand but twice as slow as using sets. Regex is about 15 times slower than sets, maybe 5 times slower than the loop.

Reply:
Have you pushed all your code? I get it quite consistently and the logic looks like it _should_ remove partial words.

The looping regex is definitely slow. The second option I proposed with the single regex is slightly faster than the original code but nowhere near the sets.

```
original 0.7861279549542814
for_version 6.605184729909524
regex_version 0.683092060033232
precalculated_regex_version 0.5726282709510997
```

The test file for these is here: https://gist.github.com/forslund/c01636a18ddef0a9bef84dae609f4511

Reply:
No, I have not re-submitted yet, but I did post the source code for my tests above in my reply to Jarbas. See class test():

Reply:
Hmm, I still have issues with that version. It works with a slight tweak:

```python
def remove_noise_preserve_order(self, phrase):
        # remove noise to produce essence
        phrase = " " + phrase + " "
        for word in self.translated_noise_words:
            mtch = " " + word + " "
            if phrase.find(mtch) > -1:
                phrase = phrase.replace(mtch, " ")  # Was phrase.replace(word, "")
        phrase = " ".join(phrase.split())
        return phrase.strip()
```

Reply:
You are correct. Thank you for this. I believe my latest commit corrects this.

Reply:
Thank you for the excellent input!